test/integration/dinosaurs_test.go

```diff
@@ -173,3 +178,94 @@ func TestDinosaurListSearch(t *testing.T) {
 	Expect(*list.Items[0].Id).To(Equal(dinosaurs[0].ID))
 }
+
+func TestUpdateDinosaurWithRacingRequests(t *testing.T) {
+	h, client := test.RegisterIntegration(t)
+
+	account := h.NewRandAccount()
+	ctx := h.NewAuthenticatedContext(account)
+
+	dino := h.NewDinosaur("Stegosaurus")
+
+	// starts 20 threads to update this dinosaur at the same time
+	threads := 20
+	var wg sync.WaitGroup
+	wg.Add(threads)
+
+	for i := 0; i < threads; i++ {
+		go func() {
+			defer wg.Done()
+			species := "Pterosaur"
+			updated, resp, err := client.DefaultApi.ApiOcmExampleServiceV1DinosaursIdPatch(ctx, dino.ID).DinosaurPatchRequest(openapi.DinosaurPatchRequest{Species: &species}).Execute()
+			Expect(err).NotTo(HaveOccurred(), "Error posting object:  %v", err)
+			Expect(resp.StatusCode).To(Equal(http.StatusOK))
+			Expect(*updated.Species).To(Equal(species), "species mismatch")
+		}()
+	}
+
+	// waits for all goroutines above to complete
+	wg.Wait()
+
+	dao := dao.NewEventDao(&h.Env().Database.SessionFactory)
+	events, err := dao.All(ctx)
+	Expect(err).NotTo(HaveOccurred(), "Error getting events:  %v", err)
+
+	updatedCount := 0
+	for _, e := range events {
+		if e.SourceID == dino.ID && e.EventType == api.UpdateEventType {
+			updatedCount = updatedCount + 1
+		}
+	}
+
+	// the dinosaur patch request is protected by the advisory lock, so there should only be one update
+	Expect(updatedCount).To(Equal(1))
+}
+
+func TestUpdateDinosaurWithRacingRequests_WithoutLock(t *testing.T) {
+	// we disable the advisory lock and try to update the dinosaurs
+	services.DisableAdvisoryLock = true
+
+	defer func() {
+		services.DisableAdvisoryLock = false
+	}()
+
+	h, client := test.RegisterIntegration(t)
+
+	account := h.NewRandAccount()
+	ctx := h.NewAuthenticatedContext(account)
+
+	dino := h.NewDinosaur("Tyrannosaurus")
+
+	// starts 20 threads to update this dinosaur at the same time
+	threads := 20
+	var wg sync.WaitGroup
+	wg.Add(threads)
+
+	for i := 0; i < threads; i++ {
+		go func() {
+			defer wg.Done()
+			species := "Triceratops"
+			updated, resp, err := client.DefaultApi.ApiOcmExampleServiceV1DinosaursIdPatch(ctx, dino.ID).DinosaurPatchRequest(openapi.DinosaurPatchRequest{Species: &species}).Execute()
+			Expect(err).NotTo(HaveOccurred(), "Error posting object:  %v", err)
+			Expect(resp.StatusCode).To(Equal(http.StatusOK))
+			Expect(*updated.Species).To(Equal(species), "species mismatch")
+		}()
+	}
+
+	// waits for all goroutines above to complete
+	wg.Wait()
+
+	dao := dao.NewEventDao(&h.Env().Database.SessionFactory)
+	events, err := dao.All(ctx)
+	Expect(err).NotTo(HaveOccurred(), "Error getting events:  %v", err)
+
+	updatedCount := 0
+	for _, e := range events {
+		if e.SourceID == dino.ID && e.EventType == api.UpdateEventType {
+			updatedCount = updatedCount + 1
+		}
+	}
+
+	// the dinosaur patch request is not protected by the advisory lock, so there should be at least one update
+	Expect(updatedCount >= 1).To(BeTrue())
```

Comment:
but with this test I think we know the _exact_ number of updates expected.

Reply:
we cannot know the exact number here, in this case, there are 20 threads try to update one dinosaur species from `Tyrannosaurus` to `Triceratops` concurrently, and because we don't have a lock here, there will be many possible outcomes here, e.g. let us consider three threads: t1, t2 and t3,

### possibility 1
t1 starts to get the dinosaur and update the dinosaur
t2 starts to get the dinosaur after the t1 is done, t2 get the updated dinosaur, it do nothing
t3 starts to get the dinosaur after the t1 is done, t3 get the updated dinosaur, it do nothing
so there are will be one update


### possibility 2
t1 starts to get the dinosaur and update the dinosaur
t2 starts to get the dinosaur when t1 performs the get but the t1 update has not yet completed, t2 will update the dinosaur again
t3 starts to get the dinosaur after the t1 is done, t3 get the updated dinosaur, it do nothing
so there are will be two updates


### possibility 3
t1 starts to get the dinosaur and update the dinosaur
t2 starts to get the dinosaur when t1 performs the get but its update has not yet completed, t2 will update the dinosaur again
t3 starts to get the dinosaur when t1 performs the get but its update has not yet completed, t3 will update the dinosaur again
so there are will be three updates

...

Reply:
ok, because this is testing _without using the locks_, then threads will have those indeterminate results, as you note. 

the other test you included that uses locks is verified as being processes exactly once.

Reply:
Let me see if I can contribute a test.  Really you want to initiate the first dinosaur rename, but have a pause between the get/update.  Then initiate a second rename, but with no pause.  Then make sure they are processed in order, where the first change is to raptor and the second is pterodactyl.

If you run this with locks the name at the end will be pterodactyl, but if you run this without locks, it will be raptor(assuming I understand the locking logic). This type of test would require changes to h.dinosaur.replace()

Reply:
yeah, I think your case is trying to simulate avoiding unrepeatable read between two transactions, the t1 get the dinosaur then pause 5s then update the dinosaur to raptor, the t2 get the dinosaur then update the dinosaur to pterodactyl without pause, if with locks, the t2 must wait for the lock, so the final result will be pterodactyl, but without lock, the t2 will update the dinosaur to pterodactyl firstly, then the t1 update the dinosaur to raptor

we can use our lock to ensure this, and I think you may introduce a pause variable only for test (like the `DisableAdvisoryLock`), such that we don't need to change the replace interface, we can only enable the pause in the test :)